// temporal-utils/dist/diff.js
function createDiffFunc(unit) {
    return (date0, date1, options) => {
        const normOptions = normalizeDiffOptions(options);
        // TODO: throw error if unit impossible for input-types?
        // like diffing years for PlainTime?
        if (normOptions.roundingMode) {
            return date0.until(date1, {
                ...normOptions,
                largestUnit: unit,
                smallestUnit: unit,
            })[unit];
        }
        const duration = date0.until(date1, {
            ...normOptions,
            largestUnit: unit,
        });
        // Time-unit helpers produce pure elapsed-time totals. This keeps Instant
        // and PlainTime valid because neither can be used as a `relativeTo` value.
        if (isTimeUnit(unit)) {
            return duration.total(unit);
        }
        // Date-bearing objects need `relativeTo` so totals account for calendar
        // units and, for ZonedDateTime, time-zone transitions. PlainYearMonth has
        // month precision, so day 1 gives Duration.total a concrete date anchor.
        const relativeTo = (!('day' in date0) && 'toPlainDate' in date0
            ? date0.toPlainDate({ day: 1 })
            : date0);
        return duration.total({ unit, relativeTo });
    };
}
// TODO: more DRY with elsewhere?
function isTimeUnit(unit) {
    return (unit === 'hours' ||
        unit === 'minutes' ||
        unit === 'seconds' ||
        unit === 'milliseconds' ||
        unit === 'microseconds' ||
        unit === 'nanoseconds');
}
export const diffYears = createDiffFunc('years');
export const diffMonths = createDiffFunc('months');
export const diffWeeks = createDiffFunc('weeks');
export const diffDays = createDiffFunc('days');
export const diffHours = createDiffFunc('hours');
export const diffMinutes = createDiffFunc('minutes');
export const diffSeconds = createDiffFunc('seconds');
export const diffMilliseconds = createDiffFunc('milliseconds');
export const diffMicroseconds = createDiffFunc('microseconds');
export const diffNanoseconds = createDiffFunc('nanoseconds');
function normalizeDiffOptions(options) {
    return typeof options === 'string' ? { roundingMode: options } : options || {};
}
