// @typespec/compiler/dist/src/core/cli/download-vsix.js
import { mkdtemp, rm, writeFile } from "fs/promises";
import os from "os";
import { createDiagnostic } from "../messages.js";
import { joinPaths } from "../path-utils.js";
import { NoTarget } from "../types.js";
const MARKETPLACE_URL = "https://marketplace.visualstudio.com";
/**
 * Download the latest `.vsix` for the given extension from the Visual Studio Marketplace
 * into a temporary directory, invoke the `install` callback with the vsix path, then clean up.
 *
 * To debug with a locally built vsix rather than pulling from the marketplace, set the
 * `TYPESPEC_DEBUG_VSIX` environment variable to the full path of the `.vsix` file.
 */
export async function downloadVsixFromMarketplace(host, extension, install) {
    const temp = await mkdtemp(joinPaths(os.tmpdir(), "typespec"));
    try {
        let vsixPath = process.env.TYPESPEC_DEBUG_VSIX;
        if (vsixPath === undefined) {
            const version = await resolveLatestVersion(extension.id);
            host.logger.trace(`Downloading ${extension.id}@${version} from the marketplace.`);
            const content = await downloadVsix(extension, version);
            vsixPath = joinPaths(temp, `${extension.name}.vsix`);
            await writeFile(vsixPath, content);
        }
        install(vsixPath);
        return [];
    }
    catch (error) {
        return [
            createDiagnostic({
                code: "vsix-download-failed",
                format: {
                    id: extension.id,
                    message: error instanceof Error ? error.message : String(error),
                },
                target: NoTarget,
            }),
        ];
    }
    finally {
        await rm(temp, { recursive: true, force: true });
    }
}
async function resolveLatestVersion(id) {
    const res = await fetch(`${MARKETPLACE_URL}/_apis/public/gallery/extensionquery`, {
        method: "POST",
        headers: {
            Accept: "application/json;api-version=3.0-preview.1",
            "Content-Type": "application/json",
        },
        // filterType 7 matches an extension by its full `<publisher>.<name>` id.
        // flags 0x1 (IncludeVersions) returns the versions ordered latest first.
        body: JSON.stringify({
            filters: [{ criteria: [{ filterType: 7, value: id }] }],
            flags: 0x1,
        }),
    });
    if (!res.ok) {
        throw new Error(`Marketplace query returned status ${res.status}.`);
    }
    const data = await res.json();
    const version = data?.results?.[0]?.extensions?.[0]?.versions?.[0]?.version;
    if (typeof version !== "string") {
        throw new Error(`Extension "${id}" was not found in the marketplace.`);
    }
    return version;
}
async function downloadVsix(extension, version) {
    const url = `${MARKETPLACE_URL}/_apis/public/gallery/publishers/${extension.publisher}/vsextensions/${extension.name}/${version}/vspackage`;
    const res = await fetch(url);
    if (!res.ok) {
        throw new Error(`Downloading vsix returned status ${res.status}.`);
    }
    return Buffer.from(await res.arrayBuffer());
}
