// @typespec/compiler/dist/src/core/features.js
import { getLocationContext } from "./helpers/location-context.js";
export const compilerFeatures = {
    "function-declarations": {
        description: "Allows use of function declarations without experimental warnings in project code.",
    },
    "auto-decorators": {
        description: "Allows use of auto decorator declarations without experimental warnings in project code.",
    },
};
export const compilerFeatureNames = Object.keys(compilerFeatures);
const compilerFeatureNameSet = new Set(compilerFeatureNames);
export function isCompilerFeatureName(feature) {
    return compilerFeatureNameSet.has(feature);
}
export function isCompilerFeatureEnabled(program, feature, target) {
    // Without a target, fall back to the root project config (e.g. global feature checks).
    if (target === undefined) {
        return program.compilerOptions.configFile?.features?.includes(feature) ?? false;
    }
    // Resolve the feature set from the package that owns the source file so that a library
    // can enable a feature for its own code via its own `tspconfig.yaml`, independently of
    // the consuming project's config.
    const context = getLocationContext(program, target);
    switch (context.type) {
        case "project":
            return program.compilerOptions.configFile?.features?.includes(feature) ?? false;
        case "library":
            return context.features?.includes(feature) ?? false;
        default:
            // compiler (standard library) and synthetic code are never gated.
            return false;
    }
}
