// @typespec/compiler/dist/src/core/cli/actions/vscode.js
import { createDiagnostic } from "../../messages.js";
import { NoTarget } from "../../types.js";
import { reportDeprecatedCommand, run } from "../utils.js";
/** Marketplace identifier of the TypeSpec VS Code extension. */
const VSCODE_EXTENSION_ID = "microsoft.typespec-vscode";
/** Documentation page describing how to install/manage the VS Code extension. */
const VSCODE_DOCS_URL = "https://typespec.io/docs/introduction/editor/vscode/";
export async function installVSCodeExtension(host, options) {
    reportDeprecatedCommand(host, "tsp code install", VSCODE_DOCS_URL);
    return runCode(host, ["--install-extension", VSCODE_EXTENSION_ID], options.insiders);
}
export async function uninstallVSCodeExtension(host, options) {
    reportDeprecatedCommand(host, "tsp code uninstall", VSCODE_DOCS_URL);
    return runCode(host, ["--uninstall-extension", VSCODE_EXTENSION_ID], options.insiders);
}
function runCode(host, codeArgs, insiders) {
    try {
        run(host, insiders ? "code-insiders" : "code", codeArgs, {
            // VS Code's CLI emits node warnings that we can't do anything about. Suppress them.
            extraEnv: { NODE_NO_WARNINGS: "1" },
            allowNotFound: true,
        });
        return [];
    }
    catch (error) {
        if (error.code === "ENOENT") {
            host.logger.trace(error.stack);
            return [
                createDiagnostic({
                    code: "vscode-in-path",
                    messageId: process.platform === "darwin" ? "osx" : "default",
                    target: NoTarget,
                }),
            ];
        }
        throw error;
    }
}
